Killing program on pod selection results in an error

```
goproxie                                                                                                                                                                                                                                          
? Choose project: <project>
? Choose proxy type: POD
? Choose cluster:  [Use arrows to move, type to filter]
> <cluster>
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x5614a8]

goroutine 1 [running]:
github.com/AckeeCZ/goproxie/internal/gcloud.GetClusterCredentials(...)
	/home/smolijar/go/src/github.com/AckeeCZ/goproxie/internal/gcloud/cli.go:39
main.main()
	/home/smolijar/go/src/github.com/AckeeCZ/goproxie/main.go:234 +0x148
```

--- a/main.go
+++ b/main.go
@@ -278,7 +278,15 @@ func main() {
 		gcloudGetClusterCredentials(projectID, cluster)
 		loadingStop()
 		namespace := readNamespace()
+		if namespace == "" {
+			fmt.Println("Could not find any GCP Clusters")
+			return
+		}
 		pod := readPod(namespace)
+		if pod == nil {
+			fmt.Printf("Could not find any K8S Pods in namespace %v", namespace)
+			return
+		}
 		remotePort := readRemotePort(pod.ContainerPorts)
 		localPort := readLocalPort(remotePort)
 		if *flags.noSave == false {

--- a/main.go
+++ b/main.go
@@ -17,11 +17,18 @@ import (
 	"github.com/briandowns/spinner"
 )
 
+var gcloudProjectsList = gcloud.ProjectsList
+var kubectlPodsList = kubectl.PodsList
+var gcloudContainerClustersList = gcloud.ContainerClustersList
+var gcloudGetClusterCredentials = gcloud.GetClusterCredentials
+var kubectlNamespacesList = kubectl.NamespacesList
+var kubectlPortForward = kubectl.PortForward
+
 func initializationCheck() {
 	// TODO
 }
 
-func readProxyType() string {
+var readProxyType = func() string {
 	proxyType := ""
 	proxyTypes := []string{"VM", "POD"}
 	if *flags.proxyType != "" {
@@ -119,32 +126,36 @@ func promptSelection(sel selectField) interface{} {
 	return pickedOption.value
 }
 
-func readProjectID() string {
-	return promptSelection(selectField{
+func readProjectID() (projectID string) {
+	// Cannot return directly, I have to accept both return values to avoid crash 🤷
+	// https://gist.github.com/smoliji/f72fe94b028125a22efa53a430ba007a
+	projectID, _ = promptSelection(selectField{
 		titleLoading: "GCP Projects",
 		titleChoose:  "GCP Project",
 		getOptions: func() (options []selectFieldOption) {
-			for _, project := range gcloud.ProjectsList() {
+			for _, project := range gcloudProjectsList() {
 				options = append(options, selectFieldOption{title: project, value: project})
 			}
 			return
 		},
 		valueTitle: *flags.project,
 	}).(string)
+	return
 }
 
-func readCluster(projectID string) *gcloud.Cluster {
-	return promptSelection(selectField{
+func readCluster(projectID string) (cluster *gcloud.Cluster) {
+	cluster, _ = promptSelection(selectField{
 		titleLoading: "Clusters",
 		titleChoose:  "Cluster",
 		getOptions: func() (options []selectFieldOption) {
-			for _, cluster := range gcloud.ContainerClustersList(projectID) {
+			for _, cluster := range gcloudContainerClustersList(projectID) {
 				options = append(options, selectFieldOption{title: cluster.Name, value: cluster})
 			}
 			return
 		},
 		valueTitle: *flags.cluster,
 	}).(*gcloud.Cluster)
+	return
 }
 
 func filterStrings(options []string, filter string) []string {
@@ -164,32 +175,34 @@ func filterStrings(options []string, filter string) []string {
 	return results
 }
 
-func readNamespace() string {
-	return promptSelection(selectField{
+func readNamespace() (namespace string) {
+	namespace, _ = promptSelection(selectField{
 		titleLoading: "K8S Namespaces",
 		titleChoose:  "K8S Namespace",
 		getOptions: func() (options []selectFieldOption) {
-			for _, namespace := range kubectl.NamespacesList() {
+			for _, namespace := range kubectlNamespacesList() {
 				options = append(options, selectFieldOption{title: namespace, value: namespace})
 			}
 			return
 		},
 		valueTitle: *flags.namespace,
 	}).(string)
+	return
 }
 
-func readPod(namespace string) *kubectl.Pod {
-	return promptSelection(selectField{
+func readPod(namespace string) (pod *kubectl.Pod) {
+	pod, _ = promptSelection(selectField{
 		titleLoading: "Pods",
 		titleChoose:  "Pod",
 		getOptions: func() (options []selectFieldOption) {
-			for _, pod := range kubectl.PodsList(namespace) {
+			for _, pod := range kubectlPodsList(namespace) {
 				options = append(options, selectFieldOption{title: pod.Name, value: pod})
 			}
 			return
 		},
 		valueTitle: *flags.pod,
 	}).(*kubectl.Pod)
+	return
 }
 
 func readLocalPort(defaultPort int) int {
@@ -211,8 +224,8 @@ func readLocalPort(defaultPort int) int {
 	return n
 }
 
-func readRemotePort(containerPorts []int) int {
-	return promptSelection(selectField{
+func readRemotePort(containerPorts []int) (port int) {
+	port, _ = promptSelection(selectField{
 		titleLoading: "Remote ports",
 		titleChoose:  "Remote port",
 		getOptions: func() (options []selectFieldOption) {
@@ -223,6 +236,7 @@ func readRemotePort(containerPorts []int) int {
 		},
 		valueTitle: *flags.remotePort,
 	}).(int)
+	return
 }
 
 func readArguments() {
@@ -249,11 +263,19 @@ func main() {
 		return
 	}
 	projectID := readProjectID()
+	if projectID == "" {
+		fmt.Println("Could not find any GCP Projects")
+		return
+	}
 	proxyType := readProxyType()
 	cluster := readCluster(projectID)
+	if cluster == nil {
+		fmt.Println("Could not find any GCP Clusters")
+		return
+	}
 	if proxyType == "POD" {
 		loadingStart("Loading Cluster credentials")
-		gcloud.GetClusterCredentials(projectID, cluster)
+		gcloudGetClusterCredentials(projectID, cluster)
 		loadingStop()
 		namespace := readNamespace()
 		pod := readPod(namespace)
@@ -262,7 +284,7 @@ func main() {
 		if *flags.noSave == false {
 			history.StorePodProxy(projectID, cluster, namespace, pod, localPort, remotePort)
 		}
-		kubectl.PortForward(pod.Name, localPort, remotePort, namespace)
+		kubectlPortForward(pod.Name, localPort, remotePort, namespace)
 	}
 
 	// fmt.Println(project_id)

--- a/main_test.go
+++ b/main_test.go
@@ -1,0 +1,231 @@
+package main
+
+import (
+	"flag"
+	"os"
+	"testing"
+
+	"github.com/AckeeCZ/goproxie/internal/gcloud"
+	"github.com/AckeeCZ/goproxie/internal/kubectl"
+)
+
+func mockGcloudProjectList(mockedProjects []string) func() {
+	originalFn := gcloudProjectsList
+	gcloudProjectsList = func() []string {
+		return mockedProjects
+	}
+	return func() {
+		gcloudProjectsList = originalFn
+	}
+}
+
+func mockKubectlPodsList(mockedPods []*kubectl.Pod) func() {
+	originalFn := kubectlPodsList
+	kubectlPodsList = func(_ string) []*kubectl.Pod {
+		return mockedPods
+	}
+	return func() {
+		kubectlPodsList = originalFn
+	}
+}
+
+func mockGcloudContainerClustersList(mockedClusters []*gcloud.Cluster) func() {
+	originalFn := gcloudContainerClustersList
+	gcloudContainerClustersList = func(_ string) []*gcloud.Cluster {
+		return mockedClusters
+	}
+	return func() {
+		gcloudContainerClustersList = originalFn
+	}
+}
+
+func mockProxyType(proxyType string) func() {
+	originalFn := readProxyType
+	readProxyType = func() string {
+		return proxyType
+	}
+	return func() {
+		readProxyType = originalFn
+	}
+}
+
+func mockKubcetlNamespacesList(namespaces []string) func() {
+	originalFn := kubectlNamespacesList
+	kubectlNamespacesList = func() []string {
+		return namespaces
+	}
+	return func() {
+		kubectlNamespacesList = originalFn
+	}
+}
+
+func mockGcloudGetClusterCredentials() func() {
+	originalFn := gcloudGetClusterCredentials
+	gcloudGetClusterCredentials = func(_ string, _ *gcloud.Cluster) {}
+	return func() {
+		gcloudGetClusterCredentials = originalFn
+	}
+}
+
+type PortforwardArgs struct {
+	podName    string
+	localPort  int
+	remotePort int
+	namespace  string
+}
+
+func mockKubectlPortForward() func() PortforwardArgs {
+	originalFn := kubectlPortForward
+	callArgs := PortforwardArgs{}
+	kubectlPortForward = func(podName string, localPort int, remotePort int, namespace string) {
+		callArgs.podName = podName
+		callArgs.localPort = localPort
+		callArgs.remotePort = remotePort
+		callArgs.namespace = namespace
+	}
+	return func() PortforwardArgs {
+		kubectlPortForward = originalFn
+		return callArgs
+	}
+}
+
+func mockAll(
+	projects []string,
+	pods []*kubectl.Pod,
+	clusters []*gcloud.Cluster,
+	proxyType string,
+	namespace []string,
+) func() {
+	unmockProjects := mockGcloudProjectList(projects)
+	unmockPods := mockKubectlPodsList(pods)
+	unmockClusters := mockGcloudContainerClustersList(clusters)
+	unmockProxyType := mockProxyType("POD")
+	unmockGetCredentials := mockGcloudGetClusterCredentials()
+	unmockNamespaces := mockKubcetlNamespacesList(namespace)
+	return func() {
+		unmockProjects()
+		unmockPods()
+		unmockClusters()
+		unmockProxyType()
+		unmockGetCredentials()
+		unmockNamespaces()
+	}
+}
+
+func resetFlags() {
+	flag.CommandLine = flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+}
+
+func TestHappyPath(t *testing.T) {
+	resetFlags()
+	unmockAll := mockAll(
+		[]string{"project-1"},
+		[]*kubectl.Pod{
+			&kubectl.Pod{Name: "pod-1", ContainerPorts: []int{1}, Containers: []string{"container-1"}},
+		},
+		[]*gcloud.Cluster{
+			&gcloud.Cluster{Name: "cluster-1", Location: "location-1"},
+		},
+		"POD",
+		[]string{"namespace-1"},
+	)
+	defer unmockAll()
+	unmockPortForward := mockKubectlPortForward()
+	// Uses single choice auto-selection
+	os.Args = []string{"goproxie", "-local_port=1234"}
+	main()
+	calledWith := unmockPortForward()
+	if calledWith.localPort != 1234 {
+		t.Errorf("Expected port-forward to be called with localPort=%v, but was called with %v", 1234, calledWith.localPort)
+	}
+	if calledWith.remotePort != 1 {
+		t.Errorf("Expected port-forward to be called with remotePort=%v, but was called with %v", 1, calledWith.remotePort)
+	}
+	if calledWith.podName != "pod-1" {
+		t.Errorf("Expected port-forward to be called with podName=%v, but was called with %v", 1, calledWith.podName)
+	}
+	if calledWith.namespace != "namespace-1" {
+		t.Errorf("Expected port-forward to be called with namespace=%v, but was called with %v", 1, calledWith.namespace)
+	}
+}
+
+func ExampleNoProjects() {
+	resetFlags()
+	unmockAll := mockAll(
+		[]string{},
+		[]*kubectl.Pod{
+			&kubectl.Pod{Name: "pod-1", ContainerPorts: []int{1}, Containers: []string{"container-1"}},
+		},
+		[]*gcloud.Cluster{
+			&gcloud.Cluster{Name: "cluster-1", Location: "location-1"},
+		},
+		"POD",
+		[]string{"namespace-1"},
+	)
+	defer unmockAll()
+	os.Args = []string{"goproxie", "-local_port=1234"}
+	main()
+	// Output: Could not find any GCP Projects
+}
+
+func ExampleNoClusters() {
+	resetFlags()
+	unmockAll := mockAll(
+		[]string{"project-1"},
+		[]*kubectl.Pod{
+			&kubectl.Pod{Name: "pod-1", ContainerPorts: []int{1}, Containers: []string{"container-1"}},
+		},
+		[]*gcloud.Cluster{},
+		"POD",
+		[]string{"namespace-1"},
+	)
+	defer unmockAll()
+	os.Args = []string{"goproxie", "-local_port=1234"}
+	main()
+	// Output:
+	// GCP Project: project-1
+	// Could not find any GCP Clusters
+}
+
+func ExampleNoNamespaces() {
+	resetFlags()
+	unmockAll := mockAll(
+		[]string{"project-1"},
+		[]*kubectl.Pod{
+			&kubectl.Pod{Name: "pod-1", ContainerPorts: []int{1}, Containers: []string{"container-1"}},
+		},
+		[]*gcloud.Cluster{
+			&gcloud.Cluster{Name: "cluster-1", Location: "location-1"},
+		},
+		"POD",
+		[]string{},
+	)
+	defer unmockAll()
+	os.Args = []string{"goproxie", "-local_port=1234"}
+	main()
+	// Output:
+	// GCP Project: project-1
+	// Cluster: cluster-1
+	// Could not find any GCP Clusters
+}
+
+func ExampleNoPods() {
+	resetFlags()
+	unmockAll := mockAll(
+		[]string{"project-1"},
+		[]*kubectl.Pod{},
+		[]*gcloud.Cluster{
+			&gcloud.Cluster{Name: "cluster-1", Location: "location-1"},
+		},
+		"POD",
+		[]string{"namespace-1"},
+	)
+	defer unmockAll()
+	os.Args = []string{"goproxie", "-local_port=1234"}
+	main()
+	// Output:
+	// GCP Project: project-1
+	// Cluster: cluster-1
+	// K8S Namespace: namespace-1
+	// Could not find any K8S Pods in namespace namespace-1
+}

